src/python/WMComponent/WorkflowUpdater/WorkflowUpdaterPoller.py

```diff
@@ -26,6 +32,160 @@
+    return files
+
+
+def extractPileupconf(tfile, fname):
+    """
+    Extract content of given file name from sandbox tar file
+    :param tfile: sandbox tar file
+    :param fname: name of the file to extract
+    :return: content of file
+    """
+    mode = tarMode(tfile, 'r')
+    with tarfile.open(tfile, mode, encoding='utf-8') as tar:
+        f = tar.extractfile(fname)
+        data = f.read()
+        # convert our data bytes into JSON object
+        return json.loads(data)
+
+
+def blockLocations(jdoc):
+    """
+    Return dict block names and their location from provided json sandbox file.
+
+    json structure is
+    {"<type: mc>": {"<blockName>": {"FileList" [], "NumberOfEvents":1, "PhEDExNodeNames": []},
+                    "<blockName>": {"FileList" [], "NumberOfEvents":1, "PhEDExNodeNames": []}, ...}
+
+    :param jdoc: JSON document
+    :return: dict {'blockName': [rses], ...}
+    """
+    bdict = {}
+    for rec in jdoc.values():
+        for key in rec.keys():
+            doc = rec[key]
+            bdict[key] = doc['PhEDExNodeNames']
+    return bdict
+
+
+def rucioBlockLocations(rucio, blocks):
+    """
+    Return dictionary in form: {"path-to-json1": jdoc1, "path-to-json2": jdoc2}
+    :param rucio: rucio client
+    :param blocks: list of blocks
+    :return: directionary
+    """
+    # rucio returns: [{"name": blockName, "replica": list(set(rses))}, ...]
+    rucioList = rucio.getReplicaInfoForBlocks(block=blocks)
+    bdict = {}
+    for item in rucioList:
+        bdict[item['name']] = item['replica']
+    return bdict
+
+
+def checkChanges(rdict, bdict):
+    """
+    Check if there is actually any pileup changes
+    :param rdict: dict of blocks locations from rucio
+    :param bdict: dict of blocks locations from pileupconf.json sandbox file
+    :return: boolean
+    """
+    # compare block names
+    if len(rdict) != len(bdict):
+        return True
+
+    # compare lcoations
+    for key in rdict.keys():
+        if sorted(rdict[key]) != sorted(bdict[key]):
+            return True
+    return False
+
+
+def updateBlockInfo(jdoc, rdict, bdict):
+    """
+    Discard given pileup names from JSON sandbox and return new dict.
+    NOTE: this function may consume lots of memory due to jdoc structure which
+    we need to parse and create new dict since we can't discard keys in place 
+    of jdoc internal dictionary, i.e.
+
+    jdoc structure is
+    {"<type: mc>": {"<block>": {"FileList" [], "NumberOfEvents":1, "PhEDExNodeNames": []},
+                    "<block>": {"FileList" [], "NumberOfEvents":1, "PhEDExNodeNames": []}, ...}
+    and we need to discard block names. They are located in internal dictionary
+    and in place pop-up of jdoc is not allowed since it will change size of internal dictionary.
+    Therefore, to overcome this obstacles we need to copy relevant pileupName key:values
+    into new dict structure and return new dict. Such operation will mostly double the
```

Comment:
No you can't, I explicitly put comment about it since the struct we have has internal dicts which you can't change in place while iterating. Please take the struct and do your homework to see that.

Reply:
No problem, I can do your homework for you as well ;) Here you go:
https://gist.github.com/amaltaro/f205bad8efad2e6039ac1a9740a84c53

Reply:
I posted my reply directly in your gist. Bottom line except different implementation it does not change anything. We may argue which implementation better but their results are the same.

Reply:
Thanks Valentin.

Reading all of this once again, including the implementation, I think I misunderstood your sentence "and in place pop-up of jdoc is not allowed since it will change size of internal dictionary". I guess you wanted to pop blocks that we do not need such that it starts decreasing memory allocation. Is that correct? If so, **maybe** it would, but I fear that until garbage collection kicks in, we are already done with the pileup updates.

> Therefore, to overcome this obstacles we need to copy relevant pileupName key:values into new dict structure and return new dict.

Yes, this is correct and it's implemented here and in the gist that I provided.

> Such operation will mostly double the memory size.

However, the sentence above is wrong. We are not `copying` data into a new structure, we are actually passing a memory reference (this is a dict/list, so assignment is by reference instead of value).

Reply:
ok, now we are in agreement, you undesrtood correctly what I meant. Regarding doubling memory size. My suspicious is that it is worse case. And, I'm not sure how internally python does organize memory in this case. In my view when you create a new object (dict in this case) whose lifestyle is independent from the one you created Python will copy the data since the later one can be modified. In other words, if we delete objectA from `jdoc` it should not ne deleted from our `newDict`. Is it done by reference count or not it is not clear to me. Said that, I removed this offensive statement `Such operation will mostly double the memory size.` to avoid further confusions.